## Processing steps of individual Chromatin binding sample (ATAC-seq, CUT&Tag, CUT&RUN, TIP-seq datasets)
# The TIP-seq dataset required additional BAM filtering step, which remove the IVT duplication from DNA fragment originated from the same T7 promoter insertion.

# NOTE: The primary data processing of these datasets was performed using analysis pipeline build by Snakemake.
# The actual running of the pipeline was done in another dorectory as the one indicated here.

# The script below contains a minor adaptation from the exact command generated by Snakemake pipeline. 
# Please take this as a guidlines when re-analyzing our datasets


# Directory and sample setup
fastq_dir="/path/to/fastq/directory"
wd="./Nr5a2_snakemake" # mock directory name. In practice, this was done outside of this repository
sample_name="NR5A2_2C_rep1"
assay_type="CUT&Tag"

bt2_index_path="/path/to/ref/genome/mm10/bowtie2_index/mm10"

# Other tools configuration
# PICARD version 2.26.4
PICARD="/path/to/picard.jar"

# Trim_galore (version 0.6.6)
trim_galore --cores 10 --paired --quality 20 --length 20 --basename $sample_name \
 --fastqc --fastqc_args "--outdir $wd/020_fastQC/" \
 --output_dir ${wd}/010_fastq_trimmed/ \
 $fastq_dir/${sample_name}_R1.fastq.gz \
 $fastq_dir/${sample_name}_R2.fastq.gz


# Mapping to reference genome (Bowtie2 version 2.4.4)
bowtie2 -p 20 -x $bt2_index_path -t -q -N 1 -L 25 -X 1000 --no-discordant --no-mixed \
 -1 ${wd}/010_fastq_trimmed/${sample_name}_val_1.fq.gz \
 -2 010_fastq_trimmed/${sample_name}_val_2.fq.gz \
 -S 030_BAM/${sample_name}_trimmed.sam

# filtering (samtools version 1.14)
# low-quality read were filtering out using flag 1804
#  (4) read unmap
#  (8) mate unmap
#  (256) not primary alignment
#  (512) read fails platform/vendor quality checks
#  (1024) read is PCR or optical duplicate
samtools view -h --threads 20 -F 1804 -q 30 -Sb ${wd}/030_BAM/${sample_name}_trimmed.sam \
 | samtools sort --threads 20 -o ${wd}/030_BAM/${sample_name}_trimmed_filt_sort.bam
rm ${wd}/030_BAM/${sample_name}_trimmed.sam
samtools stats ${wd}/030_BAM/${sample_name}_trimmed_filt_sort.bam > ${wd}/030_BAM/stats/${sample_name}_trimmed_filt_sort.stats.txt


# Detect and remove PCR duplicate (Java (JDK version 17.0.1, PICARD version 2.26.4))
java -jar -Xmx4g $PICARD MarkDuplicates I=${wd}/030_BAM/${sample_name}_trimmed_filt_sort.bam \
 REMOVE_DUPLICATES=true O=${wd}/031_BAM_rmDUP/${sample_name}_trimmed_filt_sort_rmdup.bam \
 M=031_BAM_rmDUP/${sample_name}_trimmed_filt_sort_rmdup.txt
samtools index ${wd}/031_BAM_rmDUP/${sample_name}_trimmed_filt_sort_rmdup.bam
samtools stats ${wd}/031_BAM_rmDUP/${sample_name}_trimmed_filt_sort_rmdup.bam > ${wd}/031_BAM_rmDUP/stats/${sample_name}_trimmed_filt_sort_rmdup.stats.txt


final_bam_path=${wd}/031_BAM_rmDUP/${sample_name}_trimmed_filt_sort_rmdup.bam
if [[ $assay_type=="TIP_seq" ]]; then
    # Run T7 dedup for TIP-seq
    tiptools dedup_T7_bias -@ 4 -i ${wd}/031_BAM_rmDUP/${sample_name}_trimmed_filt_sort_rmdup.bam -o ${wd}/032_BAM_rmT7DUP/${sample_name}_trimmed_filt_sort_rmT7Dup.bam
    samtools index ${wd}/032_BAM_rmT7DUP/${sample_name}_trimmed_filt_sort_rmT7Dup.bam
    final_bam_path="${wd}/032_BAM_rmT7DUP/${sample_name}_trimmed_filt_sort_rmT7Dup.bam"
fi


# Convert to bigwig format (deepTools version 3.5.1)
bamCoverage -p 10 --extendReads --binSize 1 --smoothLength 10 --normalizeUsing RPKM --effectiveGenomeSize 2652783500 \
 --bam $final_bam_path \
 -o 050_BIGWIG/${sample_name}_50bp_RPKM.bw

# Peak calling using MACS2 (version 2.2.7.1)
macs2 callpeak --format BAMPE -t $final_bam_path \
 --gsize 2652783500 --outdir 060_MACS2_PEAK_CALLING/narrow/ -n ${sample_name}_macs2

 macs2 callpeak --broad --format BAMPE -t $final_bam_path \
 --gsize 2652783500 --outdir 060_MACS2_PEAK_CALLING/broad/ -n ${sample_name}_macs2

