## Processing steps of individual RNA-seq sample
# NOTE: The primary data processing of RNA-seq was performed using RNA-seq analysis pipeline build by Snakemake.
# The actual running of the pipeline was done in another dorectory as the one indicated here.

# The script below contains a minor adaptation from the exact command generated by Snakemake pipeline. 
# Please take this as a guidlines when re-analyzing our datasets


# Reference inputs
# We used custom genomes based on GRCm38 (Ensembl release 102) with additional annotation of co-injection marker, H2B-eGFP
STAR_ref_dir="/path/to/ref/genome/mm10/STAR_GRCm38.102_ERCC92_H2BeGFP_sjdbOverhang.100bp"
ref_gtf="/path/to/ref/genome/mm10/STAR_GRCm38.102_ERCC92_H2BeGFP_sjdbOverhang.100bp/GRCm38.102_ERCC92_H2B.eGFP.gtf"
ref_gene_bed="/path/to/ref/genome/mm10/annotation/Mus_musculus.GRCm38.102_withChr.gene.bed"
# Reference of transposable element downloaded from TEtranscript tool, which is a formated version from repeatMasker
ref_TE="/path/to/ref/genome/mm10/annotation/TEtranscripts/GRCm38_Ensembl_rmsk_TE_ChrPrefix.gtf"


# Directory and sample setup
fastq_dir="/path/to/fastq/directory"
wd="./RNA_seq_snakemake" # mock directory name. In practice, this was done outside of this repository
sample_name="Nr5a2_KD" # For example

# Trim_galore (version 0.6.6)
trim_galore --cores 10 --paired --quality 10 --length 10 \
 --fastqc --fastqc_args "--outdir ./02_fastQC/" \
 --output_dir $wd/01_fastq_trimmed/ --basename ${sample_name} \
 ${fastq_dir}/WK231_S11_L001_R1_001.fastq.gz ${fastq_dir}/WK231_S11_L001_R2_001.fastq.gz


# STAR mapping (version v2.7.10a)
STAR --runThreadN 10 --readFilesCommand zcat --genomeDir ${STAR_ref_dir} \
 --limitBAMsortRAM 3017113861 \
 --readFilesIn ${wd}/01_fastq_trimmed/${sample_name}_val_1.fq.gz ${wd}/01_fastq_trimmed/${sample_name}_val_2.fq.gz \
 --outFileNamePrefix ${wd}/03_BAM/${sample_name}_ \
 --outFilterMultimapNmax 10 --outSAMtype BAM SortedByCoordinate --outReadsUnmapped None --outSJfilterReads All 


# Filtering out low quality mapping (samtools version 1.14)
# Filtering out read containing any bit from the flag 1796:
#  (4) Read unmap
#  (256) not primary alignment
#  (512) read fails platform/vendor quality checks
#  (1024) read is PCR or optical duplicate (we haven't run PICARD, so nothing should be removed here)
samtools view -h -F 1796 -q 10 ${wd}/03_BAM/${sample_name}_Aligned.sortedByCoord.out.bam \
 | samtools sort -o ${wd}/031_BAM_filt/${sample_name}_Aligned.sortedByCoord.out.filt.bam

samtools index ${wd}/031_BAM_filt/${sample_name}_Aligned.sortedByCoord.out.filt.bam

# OPTIONAL Running infer_experiment (RSeQC version 5.0.1)
infer_experiment.py --input-file ${wd}/031_BAM_filt/${sample_name}_Aligned.sortedByCoord.out.filt.bam --refgene $ref_gene_bed > ${wd}/03_BAM/stats/${sample_name}_infer_experiment_output.txt


# OPTIONAL: Making bigwig files (deepTools version 3.5.1)
bamCoverage -p 10 --binSize 10 --bam ${wd}/031_BAM_filt/${sample_name}_Aligned.sortedByCoord.out.filt.bam \
 --normalizeUsing BPM --effectiveGenomeSize 2652783500 -o ${wd}/05_BIGWIG/${sample_name}_BPM.bw


# Gene expression quantification (FeatureCounts version 2.0.1)
# Counting exonic reads in gene
featureCounts -T 1 -t exon -g gene_id -p -s 0 -O --fraction -a $ref_gtf \
-o ${wd}/07_FEATURECOUNTS/GENES_COUNT/NOTSTRANDED/${sample_name}_COUNT_genes_exon_notstranded.txt \
 ${wd}/031_BAM_filt/${sample_name}_Aligned.sortedByCoord.out.filt.bam


# TE counts (TEtranscripts version 2.2.3)
TEcount -b ${wd}/031_BAM_filt/${sample_name}_Aligned.sortedByCoord.out.filt.bam --GTF $ref_gtf --TE $ref_TE \
 --stranded no --sortByPos --mode multi --outdir ${wd}/07_TE_COUNT/NOTSTRANDED/ --project ${sample_name}

